File: apps/frontend/packages/api/src/workspaces.ts
/**
 * IAM workspace / membership / platform-role client.
 *
 * The IAM service keeps a hand-written typed client (it is not part of the
 * OpenAPI codegen). Onboarding uses the public directory + self-service apply;
 * management calls are gated server-side (super_admin / active workspace_admin) — the
 * UI only hides what the caller can't do; the API is the real boundary.
 */
import { type ApiRequestConfig, request } from "./http";

type RequestOptions = Pick<ApiRequestConfig, "skipErrorNotify">;

export type WorkspaceSummary = { id: string; tenantId: string; name: string };

export type WorkspaceAdminView = {
  tenantId: string;
  id: string;
  name: string;
  slug: string;
  ownerUserId: string;
  systemManaged: boolean;
  joinPolicy: "open" | "approval";
  memberCount: number;
  createdAt: string;
};

export type WorkspaceMemberView = {
  userId: string;
  account: string;
  displayName: string;
  email: string;
  role: "workspace_admin" | "member";
  status: "pending" | "active" | "rejected";
  reviewedBy?: string | null;
  reviewedAt?: string | null;
  rejectionReason?: string | null;
  createdAt: string;
};

export type PlatformRole = {
  id: string;
  name: string;
  description: string;
  createdAt: string;
};

const base = "/api/iam-server";

// --- Public directory + self-service ------------------------------------

/** Public, applyable workspace list ({id,name} only). No auth required. */
export async function fetchPublicWorkspaces(options?: RequestOptions): Promise<WorkspaceSummary[]> {
  return request<WorkspaceSummary[]>({
    url: `${base}/workspaces`,
    method: "GET",
    ...options,
  });
}

/** (Re)apply to join an workspace from none/rejected → pending. */
export async function applyToWorkspace(workspaceId: string): Promise<void> {
  await request<void>({
    url: `${base}/me/memberships/${encodeURIComponent(workspaceId)}/apply`,
    method: "POST",
  });
}

// --- super_admin: workspace management -------------------------------

export type CreateWorkspaceInput = {
  tenantId: string;
  name: string;
  slug: string;
  ownerUserId?: string;
  ownerAccount?: string;
  ownerEmail?: string;
  ownerPassword?: string;
  ownerDisplayName?: string;
};

export async function listWorkspacesForAdmin(options?: RequestOptions): Promise<WorkspaceAdminView[]> {
  return request<WorkspaceAdminView[]>({
    url: `${base}/workspaces/admin`,
    method: "GET",
    ...options,
  });
}

export async function createWorkspace(input: CreateWorkspaceInput): Promise<WorkspaceAdminView> {
  return request<WorkspaceAdminView>({
    url: `${base}/workspaces`,
    method: "POST",
    data: input,
  });
}

export type CreateWorkspaceAdminInput = {
  account: string;
  email: string;
  password: string;
  displayName?: string;
};

export async function createWorkspaceAdmin(
  workspaceId: string,
  input: CreateWorkspaceAdminInput,
): Promise<WorkspaceMemberView> {
  return request<WorkspaceMemberView>({
    url: `${base}/workspaces/${encodeURIComponent(workspaceId)}/admins`,
    method: "POST",
    data: input,
  });
}

export async function transferWorkspaceOwner(workspaceId: string, newOwnerUserId: string): Promise<void> {
  await request<void>({
    url: `${base}/workspaces/${encodeURIComponent(workspaceId)}/owner`,
    method: "PUT",
    data: { newOwnerUserId },
  });
}

// --- workspace_admin (or super_admin): membership review ----------------------

export async function listWorkspaceMembers(
  workspaceId: string,
  status?: "pending" | "active" | "rejected",
  options?: RequestOptions,
): Promise<WorkspaceMemberView[]> {
  return request<WorkspaceMemberView[]>({
    url: `${base}/workspaces/${encodeURIComponent(workspaceId)}/members`,
    method: "GET",
    params: status ? { status } : undefined,
    ...options,
  });
}

export async function approveMember(workspaceId: string, userId: string): Promise<void> {
  await request<void>({
    url: `${base}/workspaces/${encodeURIComponent(workspaceId)}/members/${encodeURIComponent(userId)}/approve`,
    method: "POST",
  });
}

export async function rejectMember(workspaceId: string, userId: string, reason?: string): Promise<void> {
  await request<void>({
    url: `${base}/workspaces/${encodeURIComponent(workspaceId)}/members/${encodeURIComponent(userId)}/reject`,
    method: "POST",
    data: { reason: reason ?? "" },
  });
}

export async function setMemberRole(
  workspaceId: string,
  userId: string,
  role: "workspace_admin" | "member",
): Promise<void> {
  await request<void>({
    url: `${base}/workspaces/${encodeURIComponent(workspaceId)}/members/${encodeURIComponent(userId)}/role`,
    method: "PUT",
    data: { role },
  });
}

// --- super_admin: platform roles ----------------------------------------

export async function listPlatformRoles(): Promise<PlatformRole[]> {
  return request<PlatformRole[]>({ url: `${base}/roles`, method: "GET" });
}

export async function listUserPlatformRoles(userId: string): Promise<PlatformRole[]> {
  return request<PlatformRole[]>({
    url: `${base}/users/${encodeURIComponent(userId)}/roles`,
    method: "GET",
  });
}

export async function assignUserRole(userId: string, roleId: string): Promise<void> {
  await request<void>({
    url: `${base}/users/${encodeURIComponent(userId)}/roles`,
    method: "POST",
    data: { roleId },
  });
}

export async function removeUserRole(userId: string, roleId: string): Promise<void> {
  await request<void>({
    url: `${base}/users/${encodeURIComponent(userId)}/roles/${encodeURIComponent(roleId)}`,
    method: "DELETE",
  });
}

export type Tenant = { id: string; name: string; slug: string };
export function listTenants(): Promise<Tenant[]> {
  return request<Tenant[]>({ url: `${base}/tenants`, method: "GET" });
}
export function createTenant(input: { name: string; slug: string }): Promise<Tenant> {
  return request<Tenant>({ url: `${base}/tenants`, method: "POST", data: input });
}

File: apps/backend/services/iam/internal/application/workspaces.go
package application

import (
	"context"
	"errors"
	"strings"
	"time"

	"github.com/example/monorepo/iam/internal/application/contracts"
	"github.com/example/monorepo/iam/internal/bootstrap/config"
	"github.com/example/monorepo/iam/internal/domain"
	"github.com/example/monorepo/iam/internal/infrastructure/persistence/models"
	"github.com/example/monorepo/iam/internal/infrastructure/persistence/repositories"
	"github.com/example/monorepo/iam/internal/infrastructure/security"
)

type WorkspaceService struct {
	store *repositories.Store
	cfg   config.ServerConfig
}

func NewWorkspaceService(store *repositories.Store, cfg config.ServerConfig) *WorkspaceService {
	return &WorkspaceService{store: store, cfg: cfg}
}

// ListPublic returns the applyable workspace list as {id,name} only — no member
// counts, owners, or slugs leak to anonymous callers.
func (s *WorkspaceService) ListPublic(ctx context.Context) ([]contracts.WorkspaceSummary, error) {
	workspaces, err := s.store.ListWorkspaces(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]contracts.WorkspaceSummary, 0, len(workspaces))
	for _, workspace := range workspaces {
		out = append(out, contracts.WorkspaceSummary{TenantID: workspace.TenantID, ID: workspace.ID, Name: workspace.Name})
	}
	return out, nil
}

func (s *WorkspaceService) ListForAdmin(ctx context.Context) ([]contracts.WorkspaceAdminView, error) {
	rows, err := s.store.ListWorkspacesForAdmin(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]contracts.WorkspaceAdminView, 0, len(rows))
	for _, row := range rows {
		out = append(out, contracts.WorkspaceAdminView{
			TenantID: row.TenantID, ID: row.ID, Name: row.Name, Slug: row.Slug, OwnerUserID: row.OwnerUserID,
			SystemManaged: row.SystemManaged, JoinPolicy: row.JoinPolicy,
			MemberCount: row.MemberCount, CreatedAt: row.CreatedAt.UTC().Format(time.RFC3339),
		})
	}
	return out, nil
}

// Create makes an workspace and its first active workspace_admin/owner. Exactly one of
// {existing OwnerUserID, inline owner account} must be supplied; the calling
// super_admin is not auto-joined.
func (s *WorkspaceService) Create(ctx context.Context, req contracts.CreateWorkspaceRequest, meta AuditMeta) (contracts.WorkspaceAdminView, error) {
	if _, err := s.store.TenantByID(ctx, strings.TrimSpace(req.TenantID)); err != nil {
		return contracts.WorkspaceAdminView{}, ErrInvalidWorkspace
	}
	name := strings.TrimSpace(req.Name)
	slug := domain.NormalizeSlug(req.Slug)
	if name == "" || !domain.ValidSlug(slug) {
		return contracts.WorkspaceAdminView{}, ErrInvalidWorkspace
	}
	hasExisting := strings.TrimSpace(req.OwnerUserID) != ""
	hasInline := strings.TrimSpace(req.OwnerAccount) != ""
	if hasExisting == hasInline {
		return contracts.WorkspaceAdminView{}, ErrInvalidWorkspace // exactly one owner source
	}
	now := time.Now().UTC()
	workspace := models.Workspace{
		TenantID: strings.TrimSpace(req.TenantID), ID: NewID(), Name: name, Slug: slug, JoinPolicy: "approval",
		CreatedAt: now, UpdatedAt: now,
	}
	if hasExisting {
		ownerID := strings.TrimSpace(req.OwnerUserID)
		if _, err := s.store.UserByID(ctx, ownerID); err != nil {
			return contracts.WorkspaceAdminView{}, ErrOwnerNotFound
		}
		workspace.OwnerUserID = ownerID
		if err := mutateWithAudit(ctx, s.store, auditEntry{
			Action: "workspace.create", Actor: meta.ActorUserID, Target: ownerID, Workspace: workspace.ID,
			After: map[string]any{"workspaceId": workspace.ID, "name": workspace.Name, "slug": workspace.Slug, "ownerUserId": ownerID}, Trace: meta.TraceID,
		}, func(txStore *repositories.Store) error {
			return txStore.CreateWorkspaceWithOwner(ctx, workspace, ownerID)
		}); err != nil {
			return contracts.WorkspaceAdminView{}, ErrConflict
		}
	} else {
		account := domain.NormalizeAccount(req.OwnerAccount)
		email := domain.NormalizeEmail(req.OwnerEmail)
		if !domain.ValidAccount(account) || !domain.ValidEmail(email) || len(req.OwnerPassword) < 6 {
			return contracts.WorkspaceAdminView{}, ErrInvalidWorkspace
		}
		hash, err := security.HashPassword(req.OwnerPassword)
		if err != nil {
			return contracts.WorkspaceAdminView{}, err
		}
		display := strings.TrimSpace(req.OwnerDisplayName)
		if display == "" {
			display = account
		}
		owner := models.User{
			ID:              NewID(),
			Account:         account,
			Email:           email,
			EmailNormalized: email,
			DisplayName:     display,
			Locale:          "zh-CN",
			Timezone:        "Asia/Shanghai",
			Theme:           "system",
			CreatedAt:       now,
			UpdatedAt:       now,
		}
		workspace.OwnerUserID = owner.ID
		if err := mutateWithAudit(ctx, s.store, auditEntry{
			Action: "workspace.create", Actor: meta.ActorUserID, Target: owner.ID, Workspace: workspace.ID,
			After: map[string]any{"workspaceId": workspace.ID, "name": workspace.Name, "slug": workspace.Slug, "ownerUserId": owner.ID}, Trace: meta.TraceID,
		}, func(txStore *repositories.Store) error {
			return txStore.CreateWorkspaceWithNewOwner(ctx, workspace, owner, hash, s.cfg.GuestWorkspaceID)
		}); err != nil {
			return contracts.WorkspaceAdminView{}, ErrConflict
		}
	}
	return contracts.WorkspaceAdminView{
		TenantID: workspace.TenantID, ID: workspace.ID, Name: workspace.Name, Slug: workspace.Slug, OwnerUserID: workspace.OwnerUserID,
		JoinPolicy: workspace.JoinPolicy, MemberCount: 1, CreatedAt: workspace.CreatedAt.UTC().Format(time.RFC3339),
	}, nil
}

// CreateWorkspaceAdmin creates an account and makes it an active workspace_admin of workspaceID.
func (s *WorkspaceService) CreateWorkspaceAdmin(ctx context.Context, workspaceID string, req contracts.CreateWorkspaceAdminRequest, meta AuditMeta) (contracts.WorkspaceMemberView, error) {
	if _, err := s.store.WorkspaceByID(ctx, workspaceID); err != nil {
		return contracts.WorkspaceMemberView{}, ErrWorkspaceNotFound
	}
	account := domain.NormalizeAccount(req.Account)
	email := domain.NormalizeEmail(req.Email)
	if !domain.ValidAccount(account) || !domain.ValidEmail(email) || len(req.Password) < 6 {
		return contracts.WorkspaceMemberView{}, ErrInvalidRegistration
	}
	hash, err := security.HashPassword(req.Password)
	if err != nil {
		return contracts.WorkspaceMemberView{}, err
	}
	display := strings.TrimSpace(req.DisplayName)
	if display == "" {
		display = account
	}
	now := time.Now().UTC()
	user := models.User{
		ID:              NewID(),
		Account:         account,
		Email:           email,
		EmailNormalized: email,
		DisplayName:     display,
		Locale:          "zh-CN",
		Timezone:        "Asia/Shanghai",
		Theme:           "system",
		CreatedAt:       now,
		UpdatedAt:       now,
	}
	if err := mutateWithAudit(ctx, s.store, auditEntry{
		Action: "workspace.admin.create", Actor: meta.ActorUserID, Target: user.ID, Workspace: workspaceID,
		After: map[string]any{"userId": user.ID, "account": user.Account, "role": "workspace_admin", "status": "active"}, Trace: meta.TraceID,
	}, func(txStore *repositories.Store) error {
		return txStore.CreateUserWithMembership(ctx, user, hash, s.cfg.GuestWorkspaceID, workspaceID, "workspace_admin", "active")
	}); err != nil {
		return contracts.WorkspaceMemberView{}, ErrConflict
	}
	return contracts.WorkspaceMemberView{
		UserID:      user.ID,
		Account:     user.Account,
		DisplayName: user.DisplayName,
		Email:       user.Email,
		Role:        "workspace_admin",
		Status:      "active",
		CreatedAt:   now.UTC().Format(time.RFC3339),
	}, nil
}

func (s *WorkspaceService) TransferOwner(ctx context.Context, workspaceID, newOwnerUserID string, meta AuditMeta) error {
	newOwnerUserID = strings.TrimSpace(newOwnerUserID)
	if newOwnerUserID == "" {
		return ErrInvalidWorkspace
	}
	var oldOwner string
	if workspace, err := s.store.WorkspaceByID(ctx, workspaceID); err == nil {
		oldOwner = workspace.OwnerUserID
	}
	err := mutateWithAudit(ctx, s.store, auditEntry{
		Action:    "workspace.owner.transfer",
		Actor:     meta.ActorUserID,
		Target:    newOwnerUserID,
		Workspace: workspaceID,
		Before:    map[string]any{"ownerUserId": oldOwner},
		After:     map[string]any{"ownerUserId": newOwnerUserID},
		Trace:     meta.TraceID,
	}, func(txStore *repositories.Store) error {
		if err := txStore.TransferOwner(ctx, workspaceID, newOwnerUserID); err != nil {
			return err
		}
		return txStore.RevokeUserRefreshTokens(ctx, newOwnerUserID)
	})
	return mapMembershipErr(err)
}

func (s *WorkspaceService) ListMembers(ctx context.Context, workspaceID, status string) ([]contracts.WorkspaceMemberView, error) {
	rows, err := s.store.ListWorkspaceMembers(ctx, workspaceID, strings.TrimSpace(status))
	if err != nil {
		return nil, err
	}
	out := make([]contracts.WorkspaceMemberView, 0, len(rows))
	for _, row := range rows {
		out = append(out, workspaceMemberView(row))
	}
	return out, nil
}

func (s *WorkspaceService) Approve(ctx context.Context, workspaceID, userID, reviewerID string, meta AuditMeta) error {
	beforeRole, beforeStatus, _ := s.store.MemberRoleStatus(ctx, workspaceID, userID)
	err := mutateWithAudit(ctx, s.store, auditEntry{
		Action:    "member.approve",
		Actor:     reviewerID,
		Target:    userID,
		Workspace: workspaceID,
		Before:    map[string]any{"role": beforeRole, "status": beforeStatus},
		After:     map[string]any{"status": "active"},
		Trace:     meta.TraceID,
	}, func(txStore *repositories.Store) error {
		if err := txStore.ApproveMembership(ctx, workspaceID, userID, reviewerID); err != nil {
			return err
		}
		return txStore.RevokeUserRefreshTokens(ctx, userID)
	})
	return mapMembershipErr(err)
}

func (s *WorkspaceService) Reject(ctx context.Context, workspaceID, userID, reviewerID, reason string, meta AuditMeta) error {
	reason = strings.TrimSpace(reason)
	beforeRole, beforeStatus, _ := s.store.MemberRoleStatus(ctx, workspaceID, userID)
	err := mutateWithAudit(ctx, s.store, auditEntry{
		Action:    "member.reject",
		Actor:     reviewerID,
		Target:    userID,
		Workspace: workspaceID,
		Before:    map[string]any{"role": beforeRole, "status": beforeStatus},
		After:     map[string]any{"status": "rejected"},
		Reason:    reason,
		Trace:     meta.TraceID,
	}, func(txStore *repositories.Store) error {
		if err := txStore.RejectMembership(ctx, workspaceID, userID, reviewerID, reason); err != nil {
			return err
		}
		return txStore.RevokeUserRefreshTokens(ctx, userID)
	})
	return mapMembershipErr(err)
}

func (s *WorkspaceService) SetMemberRole(ctx context.Context, workspaceID, userID, role string, meta AuditMeta) error {
	role = strings.TrimSpace(role)
	if !domain.ValidMemberRole(role) {
		return ErrInvalidRole
	}
	beforeRole, _, _ := s.store.MemberRoleStatus(ctx, workspaceID, userID)
	err := mutateWithAudit(ctx, s.store, auditEntry{
		Action:    "member.role.set",
		Actor:     meta.ActorUserID,
		Target:    userID,
		Workspace: workspaceID,
		Before:    map[string]any{"role": beforeRole},
		After:     map[string]any{"role": role},
		Trace:     meta.TraceID,
	}, func(txStore *repositories.Store) error {
		if err := txStore.SetMemberRole(ctx, workspaceID, userID, role); err != nil {
			return err
		}
		return txStore.RevokeUserRefreshTokens(ctx, userID)
	})
	return mapMembershipErr(err)
}

// Apply lets a user (re)apply to an workspace from none/rejected.
func (s *WorkspaceService) Apply(ctx context.Context, workspaceID, userID string, meta AuditMeta) error {
	workspace, err := s.store.WorkspaceByID(ctx, workspaceID)
	if err != nil {
		return ErrWorkspaceNotFound
	}
	if workspace.SystemManaged || workspace.JoinPolicy != "approval" {
		return ErrConflict
	}
	err = mutateWithAudit(ctx, s.store, auditEntry{
		Action: "member.apply", Actor: userID, Target: userID, Workspace: workspaceID,
		After: map[string]any{"status": "pending"}, Trace: meta.TraceID,
	}, func(txStore *repositories.Store) error {
		return txStore.ApplyMembership(ctx, workspaceID, userID)
	})
	return mapMembershipErr(err)
}

func workspaceMemberView(row repositories.WorkspaceMemberRow) contracts.WorkspaceMemberView {
	view := contracts.WorkspaceMemberView{
		UserID:          row.UserID,
		Account:         row.Account,
		DisplayName:     row.DisplayName,
		Email:           row.Email,
		Role:            row.Role,
		Status:          row.Status,
		ReviewedBy:      row.ReviewedBy,
		RejectionReason: row.RejectionReason,
		CreatedAt:       row.CreatedAt.UTC().Format(time.RFC3339),
	}
	if row.ReviewedAt != nil {
		formatted := row.ReviewedAt.UTC().Format(time.RFC3339)
		view.ReviewedAt = &formatted
	}
	return view
}

func mapMembershipErr(err error) error {
	switch {
	case err == nil:
		return nil
	case errors.Is(err, repositories.ErrNotFound):
		return ErrNotFound
	case errors.Is(err, repositories.ErrConflict):
		return ErrConflict
	case errors.Is(err, repositories.ErrInvariant):
		return ErrInvariant
	case errors.Is(err, repositories.ErrNotActiveMember):
		return ErrNotActiveMember
	default:
		return err
	}
}

// ErrNotFound mirrors repositories.ErrNotFound at the service boundary for router
// mapping without importing crud there.
var ErrNotFound = errors.New("not found")
